ballista/scheduler/src/state/backend/memory.rs

```diff
@@ -0,0 +1,395 @@
+            .unwrap_or_default())
+    }
+
+    async fn put(&self, keyspace: Keyspace, key: String, value: Vec<u8>) -> Result<()> {
+        let space_key = Self::get_space_key(&keyspace);
+        if !self.states.contains_key(&space_key) {
+            self.states.insert(space_key.clone(), BTreeMap::default());
+        }
+        self.states
+            .get_mut(&space_key)
+            .unwrap()
+            .value_mut()
+            .insert(key.clone(), value.clone());
+
+        // Notify subscribers
+        let full_key = format!("{}/{}", space_key, key);
+        if let Some(res) = self.subscribers.reserve(&full_key) {
+            let event = WatchEvent::Put(full_key, value);
+            res.complete(&event);
+        }
+
+        Ok(())
+    }
+
+    async fn apply_txn(&self, ops: Vec<(Operation, Keyspace, String)>) -> Result<()> {
+        let mut locks = vec![];
+        for (_, keyspace, key) in ops.iter() {
+            locks.push(self.lock(keyspace.clone(), key).await?);
+        }
+
+        for (op, keyspace, key) in ops.into_iter() {
+            match op {
+                Operation::Delete => {
+                    self.delete(keyspace, &key).await?;
+                }
+                Operation::Put(value) => {
+                    self.put(keyspace, key, value).await?;
+                }
+            };
+        }
+
+        Ok(())
+    }
+
+    async fn mv(
+        &self,
+        from_keyspace: Keyspace,
+        to_keyspace: Keyspace,
+        key: &str,
+    ) -> Result<()> {
+        let from_space_key = Self::get_space_key(&from_keyspace);
+
+        if let Some(from_space_state) = self.states.get(&from_space_key) {
+            if let Some(state) = from_space_state.value().get(key) {
+                let ops = vec![
+                    (Operation::Delete, from_keyspace, key.to_owned()),
+                    (Operation::Put(state.clone()), to_keyspace, key.to_owned()),
+                ];
+                self.apply_txn(ops).await
+            } else {
+                // TODO should this return an error?
+                warn!(
+                    "Cannot move value at {}/{}, does not exist",
+                    from_space_key, key
+                );
+                Ok(())
+            }
+        } else {
+            // TODO should this return an error?
+            warn!(
+                "Cannot move value at {}/{}, does not exist",
+                from_space_key, key
+            );
+            Ok(())
+        }
+    }
+
+    async fn lock(&self, keyspace: Keyspace, key: &str) -> Result<Box<dyn Lock>> {
+        let flat_key = Self::get_flat_key(&keyspace, key);
+        if let Some(lock) = self.locks.get_mut(&flat_key) {
+            Ok(Box::new(lock.value().clone().lock_owned().await))
+        } else {
+            let new_lock = Arc::new(Mutex::new(()));
+            self.locks.insert(flat_key, new_lock.clone());
+            Ok(Box::new(new_lock.lock_owned().await))
+        }
+    }
```

Comment:
Hmm, I think this might allow multiple locks. 

Can we use the entry API here? 

```rust
let new_lock = Arc::new(Mutex::new(()));

let lock = self.locks.entry(&flat_key).or_insert_with(|| Arc::new(Mutex::new(())));
Ok(Box::new(lock.value().clone().lock_owned().await))
```

Reply:
Good suggestion. I'll refine it.